Decimal to FloatingPoint: Avoid losing precision by splitting operation in two, more fixes and tests

carlopi opened the pull request:
I need to review, there have to be faster way that iterative, and this looks relevant enough that we have to optimize this since it can deep in CSV reading or so.

But this should have more or less the same shape as the right solution.

carlopi committed:
TryCastDecimalToFloatingPoint: Avoid losing precision by splitting operation in two

carlopi committed:
Test on decimal to double cast

carlopi committed:
Remove floating point nudging, see 70dd9e9e45eea1c1317ae7ab22cae0233a233080

Mytherin commented:
Thanks for the PR! Looking at this I don't think the decimal conversion code is the issue, though. The problem seems to me to be in the hugeint -> double conversion code (`CastBigintToFloating`).

carlopi committed:
Move from iterative to array lookup, and skip one operation

carlopi commented:
@Mytherin: I am not sure, I think the problem in this case is the third line, the cast between DECIMAL(25,5) and DOUBLE, where 9999.99999999 skips over 1000.0 (that has a valid exact double representation). We can take a look later

```
D select 99999999999999999999.99999;
┌────────────────────────────┐
│ 99999999999999999999.99999 │
│       decimal(25,5)        │
├────────────────────────────┤
│ 99999999999999999999.99999 │
└────────────────────────────┘
D select 99999999999999999999.99999::DECIMAL(25,5);
┌───────────────────────────────────────────────────┐
│ CAST(99999999999999999999.99999 AS DECIMAL(25,5)) │
│                   decimal(25,5)                   │
├───────────────────────────────────────────────────┤
│                        99999999999999999999.99999 │
└───────────────────────────────────────────────────┘
D select 99999999999999999999.99999::DECIMAL(25,5)::DOUBLE;
┌───────────────────────────────────────────────────────────────────┐
│ CAST(CAST(99999999999999999999.99999 AS DECIMAL(25,5)) AS DOUBLE) │
│                              double                               │
├───────────────────────────────────────────────────────────────────┤
│                                            1.0000000000000002e+20 │
└───────────────────────────────────────────────────────────────────┘
D select 99999999999999999999.99999::HUGEINT::DOUBLE;
┌─────────────────────────────────────────────────────────────┐
│ CAST(CAST(99999999999999999999.99999 AS HUGEINT) AS DOUBLE) │
│                           double                            │
├─────────────────────────────────────────────────────────────┤
│                                                       1e+20 │
└─────────────────────────────────────────────────────────────┘
```

I also had a chat with @lnkuiper, and I think the update fix should fix both this cast and a weird behaviour around nudging doubles (that has been introduced as a fix to a past issue, where I think this is the proper fix)

carlopi committed:
Optimize fast path where ints are represented exactly

carlopi committed:
More weird test cases

carlopi committed:
Fix also this problem in hugeint_t code pointed out in review

carlopi commented:
@Mytherin: looked a bit deeper, there was both a problem in the hugeint_t cast AND in the Decimal cast.

First part was trivial, maybe I should add a constant for that.
Second part that I did first (...) is that even correct hugeint_t to double cast will be lossy, so order of operations matters. There are probably still some corner cases in after this, I think we might want to have tests to check that:
X >= X - 1e-30 >= X - 1e-29 >= X - 1e-28 ....

Mytherin commented on src/common/operator/cast_operators.cpp:
```diff
@@ -2634,9 +2631,72 @@ bool TryCastFromDecimal::Operation(hugeint_t input, uhugeint_t &result, CastPara
+		result = Cast::Operation<SRC, DST>(input) / DST(NumericHelper::DOUBLE_POWERS_OF_TEN[scale]);
+		return true;
+	}
+	auto power_of_ten = GetPowerOfTen(input, scale);
```
Can we add a benchmark to check what the performance difference is between the two variants?

carlopi replied:
Added the benchmark, difference is quite significant (like 80x).

There have to be alternative implementations, where less expensive versions (this can be specialised since we know the exponent is 5^x * 2^x, and up to scale 13 5^x fits in 32 bits) are used.

Here the big problem is doing hugeint_t division, that is very expensive (and we do it twice since we miss a divide + modulo combined in a single operation).

Unsure whether this makes sense as it is, that is with a big performance cliff but correct results, and a faster implementation is left as TODO, or we want to stay with the current status (that has a few errors, see the added tests that currently fail) until this is properly done.

Mytherin replied:
Thanks - 80x is quite significant indeed. `Hugeint::DivMod` could be used to avoid having to do the operation twice - but that would still leave a 40x slowdown.

Mytherin reviewed (commented):
Thanks! Looks good - one request:

carlopi committed:
Add another benchmark

carlopi committed:
Fast path also in case of no division

carlopi committed:
Refactor and expand iejoin_projection_maps.test to handle correct casting

Note that we test both (value)::DECIMAL and (1 - value)::DECIMAL, and we want to see
sum to be 1. This can be expanded to more types, more consistency / rountrip tests could
be introduced, but for now fixing the bug.

carlopi committed:
Add some benchmarks on decimal to float conversions

carlopi committed:
Fix test name

carlopi committed:
Use GetDivMod

carlopi commented:
I took another look, there are a few things that can be improved performance wise in the uhugeint / hugeint division, but those are independent of this PR.

I have a branch for those, will open them independently.

Mytherin committed:
Delete .github/regression/micro_extended.csv
